dafni_cli/api/session.py

```diff
@@ -0,0 +1,547 @@
+        and UUID.
+        """
+
+        # Continue requesting the username and password for as long as the
+        # login fails to recognise them
+        login_response = None
+        while login_response is None or not login_response.was_successful():
+            username = click.prompt("Username")
+            password = click.prompt("Password", hide_input=True)
+
+            login_response = self._login(username, password)
+
+            if not login_response.was_successful():
+                click.echo(
+                    "Failed to login. Please check your username and password and try again."
+                )
+
+        self._session_data = SessionData.from_login_response(username, login_response)
+        if self._use_session_data_file:
+            self._save_session_data()
+
+        click.echo("Login Complete")
+        self.output_user_info()
+
+    def _authenticated_request(
+        self,
+        method: Literal["get", "post", "put", "patch", "delete"],
+        url: str,
+        headers: dict,
+        data: Union[dict, BinaryIO],
+        json,
+        allow_redirect: bool,
+        recursion_level: int = 0,
+    ):
+        """Performs a an authenticated request from the DAFNI API.
+
+        Args:
+            url (str): The url endpoint that is being queried
+            headers (dict): Headers to include in the request (authorisation
+                            will already be added)
+            data (dict or BinaryIO): Data to be include in the request
+            json: Any JSON serialisable object to include in the request
+            allow_redirect (bool): Flag to allow redirects during API call.
+            recursion_level (int): Used by this method to avoid infinite loop
+                                   while attempting to refresh the access token
+
+        Returns:
+            Response
+        """
+
+        response = requests.request(
+            method,
+            url=url,
+            headers={
+                "Authorization": f"Bearer {self._session_data.access_token}",
+                **headers,
+            },
+            data=data,
+            json=json,
+            allow_redirects=allow_redirect,
+            timeout=REQUESTS_TIMEOUT,
+        )
+
+        # Check for any kind of authentication error
+        if response.status_code == 403:
+            # Try again, but only once
+            if recursion_level > 1:
+                raise RuntimeError("Could not authenticate request")
+            else:
+                self._refresh_tokens()
+                response = self._authenticated_request(
+                    method,
+                    url,
+                    headers,
+                    data,
+                    json,
+                    allow_redirect,
+                    recursion_level=recursion_level + 1,
+                )
+
+        return response
+
+    # TODO have same optional flags for each function
```

Comment:
You've potentially done this now that you use the _authenticated_request method?